### ISSUE TITLE
Add a cache feature to prevent re-execution of some function in auto reload mode

### ISSUE THREAD

**Issue by jpambrun** [gradio-app/gradio#7123]
The auto-reload is very neat. Unfortunately everything is re-executed indiscriminately. This includes resources like tensorflow dataset or database connection. 

It would be nice to have a caching mechanism to would only reload a function if it was changed like streamlit does: https://docs.streamlit.io/library/api-reference/performance/st.cache_resource.

Otherwise the, if you have stateful resources, gradio hot reload isn't really suitable. Or I could be missing something and only the documentation needs tweaking?

**Comment by freddyaboulton** [comment]
Hi @jpambrun ! Thanks for the issue. You're not missing anything - it's a limitation of the `import_module` function in the standard library. I have been thinking about ways to improve this but still brainstorming.

One thing you can do in the meantime is use reload mode in a jupyter notebook. It works like how you describe, only code under the `%%blocks` cell magic is reloaded. So you can place the stateful resources outside that cell block. You may see a flicker when the demo is reloaded but that's because of jupyter, the server is not restarted.

Some more information on that here: https://www.gradio.app/guides/developing-faster-with-reload-mode#jupyter-notebook-magic

**Comment by freddyaboulton** [comment]
Hi @jpambrun -  I am working on a fix for this in #7684 . It introduces a `gr.no_reload()` context manager. Any code in that context will not be reloaded.  Can you please test it out to see if it works?  You can install that PR with the following command. The PR also has some sample code to showcase how to use `gr.no_reload()`

```
pip install https://gradio-builds.s3.amazonaws.com/e18d2583269511a252ad3e9d610efcaee09e79d3/gradio-4.21.0-py3-none-any.whl
```

If you encounter any problems please share all code and library versions needed to reproduce so I can investigate.
